### ISSUE TITLE
Rethink strategy for saving collection in DeckPicker.onStop()

### ISSUE THREAD

**Issue by marcardar** [ankidroid/Anki-Android#4214]
I've been getting reports from a tester (Nexus 5 - 6.0.1 MMB29V) indicating that note types are not being saved even though `Activity.onStop()` is being triggered.

This won't be such a problem once #4112 is released, however it does highlight a problem with the current collection-saving strategy. Essentially, you cannot be guaranteed that an `AsyncTask` triggered in `onPause()`/`onStop()` will complete.

Probably an [`IntentService`](https://developer.android.com/reference/android/app/IntentService.html) would be one correct approach.

Please see this [SO question](http://stackoverflow.com/questions/16438391/how-to-ensure-that-data-will-be-saved-before-android-kills-a-process/16438704) for more info.

**Comment by timrae** [comment]
We should probably add save() to the end of all the DeckTasks

**Comment by marcardar** [comment]
I don't know if that would solve the problem. Essentially, you need to do work on a background thread (because it involves I/O) and it is not guaranteed to be executed if you use an `AsyncTask` in `onStop()`.

**Comment by timrae** [comment]
I mean at the end of every async tasks that performs sensitive libanki operations, like we recently did in the provider.

**Comment by marcardar** [comment]
But if the task is never executed, how will that help?

**Comment by marcardar** [comment]
Oh you mean tasks other than the onStop() task?

**Comment by timrae** [comment]
If the task I'd never executed nothing needs to be saved.

**Comment by timrae** [comment]
Yes

**Comment by marcardar** [comment]
Hmm, still not sure it solves the problem, but it might be good enough. The user may back out of the screen before the task has been executed.

**Comment by marcardar** [comment]
Don't forget, AsyncTasks are queued.

**Comment by timrae** [comment]
Thanks for your input but I think I can handle it ;)

**Comment by marcardar** [comment]
> Thanks for your input but I think I can handle it ;)

That's very encouraging. I looking forward to seeing how you do ;)

**Comment by marcardar** [comment]
Unfortunately, there is still an issue here because calling an AsyncTask from Activity.onStop() is broken (in general). It either shouldn't be called at all, or should be replaced by a call to, for example, an IntentService.

**Comment by timrae** [comment]
Can you reproduce a crash or data loss?

**Comment by marcardar** [comment]
It seems to happen consistently on some devices though not on mine, so maybe not something I can reproduce. But it should be clear that it *can* happen...

**Comment by timrae** [comment]
It's not clear at all

**Comment by marcardar** [comment]
Which part?

**Comment by timrae** [comment]
That a crash or data loss could occur in the current develop branch in the case where the AsyncTask triggered by DeckPicker.onStop() fails to execute successfully.

**Comment by marcardar** [comment]
Is it clear that the background work of the AsyncTask may never be executed?

**Comment by marcardar** [comment]
From what I can see, the background work calls Collection.save(). If this is not important, then why call the AsyncTask at all? If it is important, then you should put it in an IntentService which will definitely be called.

**Comment by timrae** [comment]
> Is it clear that the background work of the AsyncTask may never be executed?

Yes. But it's not a big deal. I think there is some scheduling related stuff that gets saved there but AFAIK no data loss occurs if it doesn't. I'm not familiar with IntentService; you're probably right that it's a better approach but it's not registering on my radar of importance.

I'll reopen with low priority in case it registers on someone else's.  ;)

**Comment by timrae** [comment]
FWIW another approach is to call col.save() on every call to answer card in libanki, so that we could probably remove the call in onStop() entirely... I contemplated it but was afraid that it might add a lag for speed-reviewers so didn't proceed.

**Comment by marcardar** [comment]
Fair enough :)

At some point you'll probably want to convert your DeckTask into a DeckIntentService. They're actually very similar so it might not be much work at all.

**Comment by marcardar** [comment]
> I contemplated it but was afraid that it might add a lag for speed-reviewers so didn't proceed.

If you are calling the decktask on each answer call, then it wouldn't add any lag. That would be at least slightly safer than calling in onStop().

**Comment by timrae** [comment]
I think it can add a lag if the result doesn't come back in time for synchronization. The code for pulling cards from the decks is slightly complicated and I don't wanna mess with it just before beta release.

**Comment by timrae** [comment]
@marcardar pointed out on the forum that this could potentially cause DB corruption

https://groups.google.com/forum/#!topic/anki-android/Ee-_-y72MQs

**Comment by mikehardy** [comment]
Just linking this to the #4950 PR, as it affected some of the lifecycle contracts, and adding this to 2.9, it wasn't done in 2.6 (or 2.8) and we may not for 2.9 by I will at least look into it and see

**Comment by mikehardy** [comment]
This is a forlorn issue - I'm untagging 2.9 and it will sit again. I did take a look but I'm not sure if the lifecycle contract changes from support libs 27 changed it enough. They guarantee that onLoadFinished() will be called inside of start/stop but that doesn't affect whether code called *in* stop will execute.
